evil-jumps.el

```diff
@@ -272,46 +281,76 @@ POS defaults to point."
+              (setf (evil-jumps-struct-previous-pos target-jump-struct) (evil-jumps-struct-previous-pos source-jump-struct))
               (setf (evil-jumps-struct-idx target-jump-struct) (evil-jumps-struct-idx source-jump-struct))
               (setf (evil-jumps-struct-ring target-jump-struct) (ring-copy source-list)))))))
     ;; delete obsolete windows
-    (maphash (lambda (key val)
+    (maphash (lambda (key _val)
                (unless (member key window-list)
                  (evil--jumps-message "removing %s" key)
                  (remhash key evil--jumps-window-jumps)))
              evil--jumps-window-jumps)))
 
 (defun evil--jump-hook (&optional command)
-  "Set jump point if COMMAND has a non-nil :jump property."
+  "`pre-command-hook' for evil-jumps.
+Set jump point if COMMAND has a non-nil `:jump' property. Otherwise,
+save the current position in case the command being executed will
+change the current buffer."
   (setq command (or command this-command))
-  (when (evil-get-command-property command :jump)
-    (evil-set-jump)))
-
-(defadvice switch-to-buffer (before evil-jumps activate)
-  (evil-set-jump))
-
-(defadvice split-window-internal (before evil-jumps activate)
-  (evil-set-jump))
-
-(eval-after-load 'etags
-  '(defadvice find-tag-noselect (before evil-jumps activate)
-     (evil-set-jump)))
+  (if (evil-get-command-property command :jump)
+      (evil-set-jump)
+    (setf (evil-jumps-struct-previous-pos (evil--jumps-get-current))
+          (point-marker))))
+
+(defun evil--jump-handle-buffer-crossing ()
+  (let ((jumping-backward evil--jumps-jumping-backward))
+    (setq evil--jumps-jumping-backward nil)
+    (dolist (frame (frame-list))
+      (dolist (window (window-list frame))
+        (let* ((struct (evil--jumps-get-current window))
+               (previous-pos (evil-jumps-struct-previous-pos struct)))
+          (when previous-pos
+            (setf (evil-jumps-struct-previous-pos struct) nil)
+            (if (and
+                 ;; `evil-jump-backward' (and other backward jumping
+                 ;; commands) needs to be handled specially. When
+                 ;; jumping backward multiple times, calling
+                 ;; `evil-set-jump' is always wrong: If you jump back
+                 ;; twice and we call `evil-set-jump' after the second
+                 ;; time, we clear the forward jump list and
+                 ;; `evil--jump-forward' won't work.
+
+                 ;; The first time you jump backward, setting a jump
+                 ;; point is sometimes correct. But we don't do it
+                 ;; here because this function is called after
+                 ;; `evil--jump-backward' has updated our position in
+                 ;; the jump list so, again, `evil-set-jump' would
+                 ;; break `evil--jump-forward'.
+                 (not jumping-backward)
+                 (let ((previous-buffer (marker-buffer previous-pos)))
+                   (and previous-buffer
```

Comment:
I think jump points can also be revived from the corresponding files right, i.e. jump points are also `(PATH POS)` pairs in addition to markers? So we should also save the `buffer-file-name` of the `previous-pos` and, if its non-nil, use `find-buffer-visiting` to determine if a buffer showing the file exists. If it does, then use it. Otherwise, when there is no buffer, save the file path along with the previous position to the jump list.

I'm unsure if the current code can also set a `(PATH POS)` pair as a jump point and use it in that way. I think the `(PATH POS)` pair is only used when loading a jumplist from the file.